### ISSUE TITLE
reifyNat: the internal rep of `KnownNat` is now `Natural`

### ISSUE THREAD

**Issue by minoki** [ekmett/reflection#41]
Since GHC 8.2/base-4.10.0.0, the internal representation of `KnownNat` class has changed from `Integer` to `Natural`.  So `reifyNat` needs to be updated to reflect this change.

Note that, for many cases, `Natural` and `Integer` has the same binary representation.  They differ if the value is greater than `maxBound :: Int`, and less than or equal to `maxBound :: Word`.

Some tests:

```haskell
> reifyNat 42 (\proxy -> GHC.TypeNats.natVal proxy == 42)
True
> reifyNat (2^63-1) (\proxy -> GHC.TypeNats.natVal proxy == 2^63-1)
True
> reifyNat (2^63) (\proxy -> GHC.TypeNats.natVal proxy == 2^63) -- should be True
False
> reifyNat (2^64-1) (\proxy -> GHC.TypeNats.natVal proxy == 2^64-1) -- should be True
False
> reifyNat (2^64) (\proxy -> GHC.TypeNats.natVal proxy == 2^64)
True
```

Tested with GHC 8.6.5 + base-4.12.0.0 + reflection-2.1.4.

**Comment by RyanGlScott** [comment]
Hm, good catch. Initially, I was tempted to fix this issue like so:

```diff
 reifyNat :: forall r. Integer -> (forall (n :: Nat). KnownNat n => Proxy n -> r) -> r
-reifyNat n k = unsafeCoerce (MagicNat k :: MagicNat r) n Proxy
+reifyNat n k = unsafeCoerce (MagicNat k :: MagicNat r)
+# if MIN_VERSION_base(4,10,0)
+                             (fromInteger n :: Natural)
+# else
+                             n
+# endif
+                             Proxy
```

That fixes your test cases, but it does pose an interesting question with how `reifyNat` should behave with respect to negative `Integer`s. In today's `reflection`, if you try to reify a negative `Integer`, then it will silently underflow:

```
λ> reifyNat (-1) reflect
18446744073709551615
```

However, I make the change above, then on `base-4.10` or later, attempting the same thing will result in an exception:

```
λ> reifyNat (-1) reflect
*** Exception: arithmetic underflow
```

How should we resolve this difference? I can think of a couple approaches:

1. Ensure that `reifyNat` always throws an exception when its argument is negative, regardless of which version of `base` is used.
2. Keep the implementation above, but document in the Haddocks for `reifyNat` that its behavior on negative `Integer`s is undefined.

Thoughts?

**Comment by minoki** [comment]
I'd prefer 1, because it would be consistent across `base` versions, and, safer.

**Comment by RyanGlScott** [comment]
Very well, I've implemented this in #42.
